src/interface/meshblock_data.cpp

```diff
@@ -199,357 +214,162 @@ std::shared_ptr<MeshBlockData<T>> MeshBlockData<T>::SparseSlice(int id) {
-MeshBlockData<T>::PackVariablesAndFluxes(const std::vector<MetadataFlag> &flags,
-                                         PackIndexMap &vmap,
-                                         vpack_types::StringPair &key) {
-  std::vector<std::string> vnams;
-  vpack_types::VarList<T> vars = MakeList_(flags, vnams);
-  return PackVariablesAndFluxesHelper_(vnams, vnams, vars, vars, vmap, key);
-}
-template <typename T>
-VariableFluxPack<T>
-MeshBlockData<T>::PackVariablesAndFluxes(const std::vector<MetadataFlag> &flags,
-                                         PackIndexMap &vmap) {
-  vpack_types::StringPair key;
-  return PackVariablesAndFluxes(flags, vmap, key);
-}
-template <typename T>
-VariableFluxPack<T>
-MeshBlockData<T>::PackVariablesAndFluxes(const std::vector<MetadataFlag> &flags,
-                                         vpack_types::StringPair &key) {
-  PackIndexMap vmap;
-  return PackVariablesAndFluxes(flags, vmap, key);
-}
-template <typename T>
-VariableFluxPack<T>
-MeshBlockData<T>::PackVariablesAndFluxes(const std::vector<MetadataFlag> &flags) {
-  PackIndexMap vmap;
-  vpack_types::StringPair key;
-  return PackVariablesAndFluxes(flags, vmap, key);
+  return itr->second;
 }
 
 /// This is a helper function that queries the cache for the given pack.
-/// The strings are the keys and the lists are the values.
+/// The strings are the key and the lists are the values.
 /// Inputs:
-/// var_names = vector of names of variables to pack
 /// vars = forward list of shared pointers of vars to pack
-/// Outputs:
-/// vmap = std::map from names to std::pairs of indices
-///        indices are the locations in the outer Kokkos::view of the pack
-///        indices represent inclusive bounds for, e.g., a sparse or tensor-valued
-///        variable.
+/// coarse = whether to use coarse pack map or not
+/// Returns:
+/// A VarMetaPack<T> that contains the actual VariablePack, the PackIndexMap, and the key
 template <typename T>
-VariablePack<T>
-MeshBlockData<T>::PackVariablesHelper_(const std::vector<std::string> &names,
-                                       const vpack_types::VarList<T> &vars,
-                                       PackIndexMap &vmap, const bool coarse) {
+const VarMetaPack<T> &MeshBlockData<T>::PackListedVariables(const VarLabelList &var_list,
+                                                            bool coarse) {
+  const auto &key = var_list.labels();
   auto &packmap = coarse ? coarseVarPackMap_ : varPackMap_;
-  auto kvpair = packmap.find(names);
-  if (kvpair == packmap.end()) {
-    auto pack = MakePack<T>(vars, &vmap, coarse);
-    PackIndxPair<T> value;
-    value.pack = pack;
-    value.map = vmap;
-    packmap[names] = value;
-    return pack;
+
+  auto itr = packmap.find(key);
+  if (itr == packmap.end()) {
+    VarMetaPack<T> new_item;
+    new_item.pack = MakePack<T>(var_list.vars(), coarse, &new_item.map);
+    itr = packmap.insert({key, new_item}).first;
+    itr->second.key = &itr->first;
```

Comment:
I'm not sure I fully follow the last two lines.
`itr` points to the new element in the map and afterwards the it's own key (which are the `var_list.labels()` is also set as the key to the pack itself?

Reply:
This has been removed now after going back to the old packing interface, but yes, the `key` of the `MetaPack` is just the key that is being used in the `packmap`. To avoid an unnecessary copy, it was just a pointer to the key.